### ISSUE TITLE
LCOM4

### ISSUE THREAD

**Issue by yegor256** [yegor256/jpeek#8]
Let's implement LCOM4.

Hitz M., Montazeri B.: Measuring Coupling and Cohesion In Object-Oriented Systems. Proc. Int. Symposium on Applied Corporate Computing, Oct. 25-27, Monterrey, Mexico, 75-76, 197, 78-84. 

http://www.isys.uni-klu.ac.at/PDF/1995-0043-MHBM.pdf

**Comment by memoyil** [comment]
@yegor256 I will look at it

**Comment by adilek** [comment]
Hi.

As you already working on it, I would like to mention something about LCOM4.
Including/excluding constructors plays a significant role and makes some anomalies.

Neither Hitz M., Montazeri B nor Chidamber did not mention about technical details like need to include constructors or not.

Lets see the following example:

```
public class MyClass {
    int var1;
    int var2;
    int var3;
    int var4;

    public MyClass(int var1, int var2, int var3, int var4) {
        this.var1 = var1;
        this.var2 = var2;
        this.var3 = var3;
        this.var4 = var4;
    }

    public int getVar1() {
        return var1;
    }

    public int getVar2() {
        return var2;
    }

    public int getVar3() {
        return var3;
    }

    public int getVar4() {
        return var4;
    }

}
```

If constructor is not included then LCOM4 will be 4. Cohesiveness is fine with this class.
If we include the constructors then LCOM will be 1. If look at it as a graph they all are connected and we have only one connected component.

If we change the same code to this:

```
public class MyClass {
    int var1;
    int var2;
    int var3;
    int var4;

    public MyClass(int var1, int var2, int var3, int var4) {
        init(var1, var2, var3, var4);
    }

    public void init(int var1, int var2, int var3, int var4) {
        this.var1 = var1;
        this.var2 = var2;
        this.var3 = var3;
        this.var4 = var4;
    }

    public int getVar1() {
        return var1;
    }

    public int getVar2() {
        return var2;
    }

    public int getVar3() {
        return var3;
    }

    public int getVar4() {
        return var4;
    }

}

```
Now it shows LCOM 1. The behaviour and cohesivness was not changed. But LCOM changed. This seems incorrect behaviour depending on constructors.

In case of including constructors we may get another problem with default constructors. For every constuctor like this:
```
public MyClass(Context ctx) {
   super(ctx);
}
```
we will get +1 for LCOM4. As you know in practice there are a lot of such examples.

As a solution in my opinion it would be good to have 2 LCOM4. One with constructors one without.

Thanks.

**Comment by adilek** [comment]
BTW, If you did start yet. I can take care of LCOM4.

**Comment by memoyil** [comment]
@adilek  Sorry for the late response, I was busy with other things. I am in the middle of lcom4. You can look at other issues I think.

**Comment by 0crat** [comment]
Job [#8](https://github.com/yegor256/jpeek/issues/8) is now in scope, role is `DEV`

**Comment by 0crat** [comment]
Bug was reported, see [§29](http://datum.zerocracy.com/pages/policy.html#29): +15 points just awarded to [@yegor256](http://www.0crat.com/u/yegor256), total is [+7150](http://www.0crat.com/u/yegor256)

**Comment by 0crat** [comment]
The job [#8](https://github.com/yegor256/jpeek/issues/8) assigned to @icok[/z](https://www.0crat.com/u/icok). The budget is 30 minutes, see [§4](http://datum.zerocracy.com/pages/policy.html#4). Please, read [§8](http://datum.zerocracy.com/pages/policy.html#8) and [§9](http://datum.zerocracy.com/pages/policy.html#9). If the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html).

**Comment by 0crat** [comment]
The job [#8](https://github.com/yegor256/jpeek/issues/8) assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil). The budget is 30 minutes, see [§4](http://datum.zerocracy.com/pages/policy.html#4). Please, read [§8](http://datum.zerocracy.com/pages/policy.html#8) and [§9](http://datum.zerocracy.com/pages/policy.html#9). If the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html).

**Comment by amihaiemil** [comment]
@0crat waiting for https://github.com/yegor256/jpeek/issues/132

**Comment by 0crat** [comment]
There is an unrecoverable failure on my side. Please, submit it [here](https://github.com/zerocracy/farm/issues):

```
com.amazonaws.SdkClientException: Unable to execute HTTP request: null
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleRetryableException(AmazonHttpClient.java:1113)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeHelper(AmazonHttpClient.java:1063)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.doExecute(AmazonHttpClient.java:743)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeWithTimer(AmazonHttpClient.java:717)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.execute(AmazonHttpClient.java:699)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.access$500(AmazonHttpClient.java:667)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutionBuilderImpl.execute(AmazonHttpClient.java:649)
	at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:513)
	at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:4247)
	at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:4194)
	at com...
```

[1.0-SNAPSHOT](https://github.com/zerocracy/farm/tree/1.0-SNAPSHOT) /cc @yegor256

**Comment by amihaiemil** [comment]
@0crat waiting for #132 (here)

**Comment by 0crat** [comment]
> @0crat waiting for #132 (here) ([here](https://github.com/yegor256/jpeek/issues/8#issuecomment-361849149))

@amihaiemil The impediment for [#8](https://github.com/yegor256/jpeek/issues/8) was registered successfully by @amihaiemil[/z](https://www.0crat.com/u/amihaiemil)

**Comment by amihaiemil** [comment]
@0crat refuse, assign someone else

**Comment by 0crat** [comment]
> @0crat refuse, assign someone else ([here](https://github.com/yegor256/jpeek/issues/8#issuecomment-363784392))

@amihaiemil Can't understand "refuse,", try one of these:
  - `assign` Assign a performer to this issue
  - `hello` Just say hello
  - `in` Register this issue in scope (WBS)
  - `out` Remove this issue from scope (WBS)
  - `resign` Resign from this issue
  - `status` Check the status of the job
  - `version` Print current version of the bot
  - `waiting` Register an impediment for a job

**Comment by amihaiemil** [comment]
@0crat resign

**Comment by 0crat** [comment]
> @0crat resign ([here](https://github.com/yegor256/jpeek/issues/8#issuecomment-363784598))

@amihaiemil The user @amihaiemil[/z](https://www.0crat.com/u/amihaiemil) resigned from [#8](https://github.com/yegor256/jpeek/issues/8), please stop working

**Comment by 0crat** [comment]
Tasks refusal is discouraged, see [§6](http://datum.zerocracy.com/pages/policy.html#6): -15 points just awarded to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), total is [+345](https://www.0crat.com/u/amihaiemil)

**Comment by 0crat** [comment]
The job [#8](https://github.com/yegor256/jpeek/issues/8) assigned to @nikialeksey[/z](https://www.0crat.com/u/nikialeksey). The budget is 30 minutes, see [§4](http://datum.zerocracy.com/pages/policy.html#4). Please, read [§8](http://datum.zerocracy.com/pages/policy.html#8) and [§9](http://datum.zerocracy.com/pages/policy.html#9). If the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html).

**Comment by 0crat** [comment]
@nikialeksey[/z](https://www.0crat.com/u/nikialeksey) this job was assigned to you 5 days ago. It will be taken away from you soon, unless you close it, see [§8](http://datum.zerocracy.com/pages/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by 0crat** [comment]
@nikialeksey[/z](https://www.0crat.com/u/nikialeksey) this job was assigned to you 8 days ago. It will be taken away from you soon, unless you close it, see [§8](http://datum.zerocracy.com/pages/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by nikialeksey** [comment]
@0crat  waiting

**Comment by 0crat** [comment]
> @0crat waiting ([here](https://github.com/yegor256/jpeek/issues/8#issuecomment-366532435))

@nikialeksey The impediment for [#8](https://github.com/yegor256/jpeek/issues/8) was registered successfully by @nikialeksey[/z](https://www.0crat.com/u/nikialeksey)

**Comment by nikialeksey** [comment]
@0crat refuse

**Comment by 0crat** [comment]
The job [#8](https://github.com/yegor256/jpeek/issues/8) assigned to @llorllale[/z](https://www.0crat.com/u/llorllale), here is [why](https://www.0crat.com/footprint/C7JGJ00DP/501346535). The budget is 30 minutes, see [§4](http://datum.zerocracy.com/pages/policy.html#4). Please, read [§8](http://datum.zerocracy.com/pages/policy.html#8) and [§9](http://datum.zerocracy.com/pages/policy.html#9). If the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html).

**Comment by llorllale** [comment]
@adilek can you please edit your comments so that they're addressed to someone in particular?

**Comment by 0crat** [comment]
> @0crat refuse ([here](https://github.com/yegor256/jpeek/issues/8#issuecomment-366536251))

@nikialeksey The job [#8](https://github.com/yegor256/jpeek/issues/8) is assigned to @llorllale[/z](https://www.0crat.com/u/llorllale), you cannot resign, since you are not a `PO` or `ARC`

**Comment by llorllale** [comment]
@yegor256 
Initial implementation was merged in PR #206 while leaving two puzzles to finish work on this metric:
* puzzle #215 due to impediment on #156 
* puzzle #216 in order to finish implementing the tests

Can we close this?

**Comment by yegor256** [comment]
@llorllale thanks!

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job, as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by ypshenychka** [comment]
@adilek According to our [QA Rules](http://www.zerocracy.com/policy.html#42):
>Messages in a job [always](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html#4-avoid-noiseaddress-your-comments) start with a name of a user they are addressed to.

Please correct your messages by indicating an addressee in the beginning.

**Comment by ypshenychka** [comment]
@adilek ping

**Comment by 0crat** [comment]
The user @llorllale[/z](https://www.0crat.com/u/llorllale) resigned from [#8](https://github.com/yegor256/jpeek/issues/8), please stop working. Reason for job resignation: It is older than 10 days, see [§8](http://www.zerocracy.com/policy.html#8)

**Comment by 0crat** [comment]
Resigned on delay, see [§8](http://www.zerocracy.com/policy.html#8): -30 points just awarded to @llorllale[/z](https://www.0crat.com/u/llorllale)

**Comment by ypshenychka** [comment]
@adilek Please consider my [remark](https://github.com/yegor256/jpeek/issues/8#issuecomment-368787292).

**Comment by ypshenychka** [comment]
@adilek ping

**Comment by ypshenychka** [comment]
@adilek ping

**Comment by 0crat** [comment]
The job [#8](https://github.com/yegor256/jpeek/issues/8) is now out of scope

**Comment by ypshenychka** [comment]
@adilek ping

**Comment by ypshenychka** [comment]
@0crat quality acceptable

**Comment by 0crat** [comment]
> @0crat quality acceptable ([here](https://github.com/yegor256/jpeek/issues/8#issuecomment-380006248))

@ypshenychka You @ypshenychka[/z](https://www.0crat.com/u/ypshenychka) need to have one of these roles in order to do "Complete `QA` review": `ARC`, `QA`; I'm sorry to say this, but at the moment you've got no roles in this project

**Comment by 0pdd** [comment]
@yegor256 3 puzzles [#215](https://github.com/yegor256/jpeek/issues/215), [#294](https://github.com/yegor256/jpeek/issues/294), [#295](https://github.com/yegor256/jpeek/issues/295) are still not solved; solved: [#216](https://github.com/yegor256/jpeek/issues/216).

**Comment by 0pdd** [comment]
@yegor256 3 puzzles [#294](https://github.com/yegor256/jpeek/issues/294), [#295](https://github.com/yegor256/jpeek/issues/295), [#310](https://github.com/yegor256/jpeek/issues/310) are still not solved; solved: [#215](https://github.com/yegor256/jpeek/issues/215), [#216](https://github.com/yegor256/jpeek/issues/216).

**Comment by 0pdd** [comment]
@yegor256 2 puzzles [#295](https://github.com/yegor256/jpeek/issues/295), [#310](https://github.com/yegor256/jpeek/issues/310) are still not solved; solved: [#215](https://github.com/yegor256/jpeek/issues/215), [#216](https://github.com/yegor256/jpeek/issues/216), [#294](https://github.com/yegor256/jpeek/issues/294).

**Comment by 0pdd** [comment]
@yegor256 2 puzzles [#295](https://github.com/yegor256/jpeek/issues/295), [#415](https://github.com/yegor256/jpeek/issues/415) are still not solved; solved: [#215](https://github.com/yegor256/jpeek/issues/215), [#216](https://github.com/yegor256/jpeek/issues/216), [#294](https://github.com/yegor256/jpeek/issues/294), [#310](https://github.com/yegor256/jpeek/issues/310).

**Comment by 0pdd** [comment]
@yegor256 2 puzzles [#295](https://github.com/yegor256/jpeek/issues/295), [#431](https://github.com/yegor256/jpeek/issues/431) are still not solved; solved: [#215](https://github.com/yegor256/jpeek/issues/215), [#216](https://github.com/yegor256/jpeek/issues/216), [#294](https://github.com/yegor256/jpeek/issues/294), [#310](https://github.com/yegor256/jpeek/issues/310), [#415](https://github.com/yegor256/jpeek/issues/415).
